codec-http2/src/main/java/io/netty/handler/codec/http2/ClearTextHttp2ServerDispatcher.java

```diff
@@ -0,0 +1,84 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandler;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.AsciiString;
+import io.netty.util.internal.UnstableApi;
+
+import java.util.List;
+
+import static io.netty.buffer.Unpooled.*;
+
+/**
+ * Provide how starting http2 protocol by h2c (Upgrade) or h2 (Prior Knowledge) by peaking the
+ * first bytes is prior knowledge or not, see {@link #PRI_KNOWLEDGE_FIRST_BYTES}
+ * The dispatcher should be the first {@link ChannelInboundHandler} in the
+ * {@link io.netty.channel.ChannelPipeline}
+ * The msg from {@link #channelRead(ChannelHandlerContext, Object)} must be {@link ByteBuf}
+ */
+@UnstableApi
+public abstract class ClearTextHttp2ServerDispatcher extends ByteToMessageDecoder {
+    private static final AsciiString PRI_KNOWLEDGE_MAGIC = AsciiString.of("PRI");
+    private static final int PRI_KNOWLEDGE_PEAK_LENGTH = PRI_KNOWLEDGE_MAGIC.length();
+    private static final ByteBuf PRI_KNOWLEDGE_FIRST_BYTES = unreleasableBuffer(
+            directBuffer(PRI_KNOWLEDGE_PEAK_LENGTH).writeBytes(PRI_KNOWLEDGE_MAGIC.array()));
+
+    /**
+     * Peak the first {@link #PRI_KNOWLEDGE_PEAK_LENGTH} bytes of msg, and verify that
+     * if it matches to {@link #PRI_KNOWLEDGE_FIRST_BYTES} then assume client want to
+     * start HTTP/2 connection with prior knowledge, call {@link #configurePriorKnowledge(ChannelHandlerContext)}
+     * to configure pipeline for handling HTTP/2 connection. If it didn't match, then
+     * call {@link #configureUpgrade(ChannelHandlerContext)} for HTTP/2 upgrade
+     *
+     * @throws IllegalStateException
+     *     if {@code !(msg instanceof io.netty.ByteBuf}
+     *     or if {@code msg.readableBytes() < PRI_KNOWLEDGE_PEAK_LENGTH}
+     */
+    @Override
+    public void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (in.readableBytes() < PRI_KNOWLEDGE_PEAK_LENGTH) {
+            return;
+        }
+
+        if (ByteBufUtil.equals(PRI_KNOWLEDGE_FIRST_BYTES, 0,
+                               in, 0, PRI_KNOWLEDGE_PEAK_LENGTH)) {
+            configurePriorKnowledge(ctx);
+        } else {
+            configureUpgrade(ctx);
+        }
+
+        out.add(in.readBytes(in.readableBytes()));
+
+        ctx.pipeline().remove(this);
+    }
+
+    /**
+     * Invoked when dispatcher knows that the client want to start HTTP/2 by upgrade.
+     * Implement this method to configure your pipeline for handling HTTP -> HTTP/2 upgrade.
+     */
+    protected abstract void configureUpgrade(ChannelHandlerContext ctx);
```

Comment:
I'm not sure that I understand about your concern.
Here is my understand about your concern, correct me if I'm wrong.

The ```HttpServerUpgradeHandler``` should be added at the start point. Like 
```
channel.pipeline().addLast(
    new CleartextHttp2ServerUpgradeHandler(),
    httpCodec, httpServerUpgradeHandler);
```
So when the client start HTTP/2  by upgrade, CleartextHttp2ServerUpgradeHandler should just remove itself from pipeline.
When client start HTTP/2 by prior knowledge, we configure pipeline for prior knowledge.

I think that what you had mention at [5857#comment](https://github.com/netty/netty/issues/5857#issue-178968404) and how finagle([code](https://github.com/twitter/finagle/blob/b46a3f96002da7a7d50d197b8907d968fc4640ca/finagle-http2/src/main/scala/com/twitter/finagle/http2/Http2CleartextServerInitializer.scala#L88-L92)) do!

I will say that your approach and mine had different idea and I think both ways make sense,
your approach assume that client will start HTTP/2 upgrade or HTTP/1.1 by default, so it config its pipeline for that, and when found client start HTTP/2 by prior knowledge, it removed handlers configured before.

And my approach is I don't assume what client want to do, I peek the first bytes and configured the pipeline when we know what client wants to do.
So that's why I named the class as a Dispatcher (I agreed it's not a good name), to dispatch ```HTTP/2 via prior knowledge``` or ```HTTP/1 + HTTP/2 via upgrade```.

I think both ways are all good and had its reason.
If we choose your way after discussion, I think I will close this one, and create another one and steal some of your code from finagle :-) , or maybe you would like to have one by yourself.